zephyr/lib/regions_mm.c

```diff
@@ -1,54 +1,563 @@
+
+			bitaray_bitfield = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				sizeof(uint32_t) * bitaray_size);
+
+			if (bitaray_bitfield == NULL) {
+				rfree(new_heap);
+				rfree(new_allocator);
+				rfree(allocators_bitaray);
+				return NULL;
+			}
+
+			allocators_bitaray->num_bits =
+				cfg->block_bundles_table[i].number_of_blocks;
+			allocators_bitaray->num_bundles = bitaray_size;
+			allocators_bitaray->bundles = bitaray_bitfield;
+
+			new_allocator->num_blocks =
+				cfg->block_bundles_table[i].number_of_blocks;
+			new_allocator->blk_sz_shift =
+				ilog2(cfg->block_bundles_table[i].block_size);
+			new_allocator->buffer =
+				(uint8_t *)new_heap->virtual_region->addr +
+					offset;
+			new_allocator->bitmap = allocators_bitaray;
+
+			offset += cfg->block_bundles_table[i].number_of_blocks *
+				cfg->block_bundles_table[i].block_size;
+			new_heap->physical_blocks_allocators[i] = new_allocator;
+		}
+	}
+
+	new_heap->span = span;
+
+	/* Save the new heap pointer to a linked list */
+	list_item_append(&vm_heaps, &new_heap->node);
+
+	return new_heap;
+fail:
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES; i++) {
+		struct sys_mem_blocks *allocator =
+			new_heap->physical_blocks_allocators[i];
+		if (!allocator) {
+			if (!allocator->bitmap)
+				rfree(allocator->bitmap);
+			rfree(allocator);
+		}
+	}
+	if (!new_heap)
+		rfree(new_heap);
+
+	return NULL;
+}
+
+/**
+ * @brief Alloc function
+ *
+ * Allocates memory on heap from provided heap pointer.
+ * Check if we need to map physical memory for given allocation
+ * and do it when nesscesary.
+ * @param heap pointer to a heap that will be used for allocation.
+ * @param alloc_size Size in bytes to allocate.
+ *
+ * @retval ptr to a new heap if successful
+ * @retval NULL on allocation failure
+ */
+void *vmh_alloc(struct virtual_memory_heap *heap, uint32_t alloc_size)
+{
+	void *ptr = NULL;
+	size_t mem_block_iterator;
+	uint32_t blck_count;
+
+	/* Error code we will gather when allocating on physical block
+	 * allocators.
+	 * Loop will continiue as long as there are blocks to iterate on and
+	 * if allocation keeps returning error code.
+	 * When alloc succeeded it will return 0 on allocation and we break
+	 * form loop and continiue.
+	 */
+	int allocation_error_code = -ENOMEM;
+	struct vmh_active_allocation *new_allocation =
+		rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			sizeof(struct vmh_active_allocation));
+
+	if (new_allocation == NULL)
+		return NULL;
+
+	list_init(&new_allocation->node);
+
+	/* If span is on for heap allocation will be made on first available block
+	 * series on given block bundle.
+	 * For span off allocation will only happened on first free block
+	 * on first allocator that can accommodate requested size.
+	 */
+	for (mem_block_iterator = 0; mem_block_iterator < MAXIMUM_BLOCK_SIZES;
+		mem_block_iterator++) {
+		/* calculate block count needed to allocate for current
+		 * mem_block.
+		 */
+		uint32_t blck_size =
```

Comment:
`blck` seems weird: either use one more character and make it `block` or use the usual abbreviation `blk`